wgsl/index.bs

```diff
@@ -6981,13 +6979,326 @@ See [[#statements]] and [[#function-calls]].
+</div>
+
+In more detail, here is what we compute for each function:
+
+* A tag about the control-flow in which the function can be called: one of {AlwaysMustBeUniform, MustBeUniformForControlFlowAfter}
+* A tag for each parameter of the function: one of {AlwaysMustBeUniform, MustBeUniformForControlFlowAfter, MustBeUniformForReturnValue, NoRestriction}
+* A tag for the function as a whole: one of {ControlFlowCannotBeUniformAfter, ReturnValueCannotBeUniform, NoRestriction}
+
+Here is the algorithm for computing this data for a given function:
+
+* Create nodes called "MustBeUniform", "CannotBeUniform", "CF_start", "CF_return", and if the function is non-void a node called "Value_return"
+* Create one node for each parameter of the function which we'll call "arg_i"
+* Walk over the syntax of the function, adding nodes and edges to the graph following the rules of the next section, using CF_start as the starting control-flow for the function's body.
+* Look at which nodes are reachable from "MustBeUniform"
+    * If this set includes the node "CannotBeUniform", then reject the program.
+    * If this set includes "CF_start", then the control-flow in which the function can be called is AlwaysMustBeUniform.
+    * Otherwise it is MustBeUniformForControlFlowAfter
+    * For each "arg_i" in this set, the corresponding parameter is AlwaysMustBeUniform
+    * Remove from the graph all nodes that have been visited
+* Look at which nodes are reachable from "CF_return"
+    * If this set includes "CannotBeUniform", then the function is ControlFlowCannotBeUniformAfter
+    * For each "arg_i" in this set, the corresponding parameter is MustBeUniformForControlFlowAfter
+    * Remove from the graph all nodes that have been visited
+* If "Value_return" exists, look at which nodes are reachable from it
+    * If this set includes "CannotBeUniform", then the function is ReturnValueCannotBeUniform
+    * For each "arg_i" in this set, the corresponding parameter is MustBeUniformForReturnValue
+* If the function has not been assigned a tag, then it is NoRestriction.
+* For each parameter, if it has not been assigned a tag, then it is NoRestriction.
+
+Note: The entire graph can be destroyed at this point. The tags listed above are all that we need to remember to analyze callers of this function.
+
+### Uniformity rules for statements ### {#uniformity-statements}
+
+In order to properly deal with reconvergence we must know when there is a single way for control to flow out of a statement.
+This is achieved by following the behaviors analysis (see [[#behaviors]]).
+
+If the set of behaviors for a statement is {Next}, this means that we either did not diverge within the statement, or we reconverged, and we pick the node corresponding to control flow at the start of the statement as the node corresponding to control flow at the exit of the statement, overriding what the other rules below say about the resulting control-flow node.
```

Comment:
Thinking about this change some more. Imagine:

```
fn foo(i : i32) {
  loop {
    if (i < 5) {
      break;
    } else {
      break;
    }
    continuing {
      textureSample(...);
    }
  }
}
```

The whole continuing block is dead, but currently the behaviour analysis doesn't consider that to be the case. If dead code is changed to no longer be an error then this analysis should probably skip dead code entirely (and loop behaviours should be updated around continuing blocks).

If the analysis doesn't identify the continuing block as dead, then it seems like there would be an edge from the textureSample statement through the if and would mark parameter `i` as needing to be uniform (which isn't true).

Reply:
More thoughts related to the singleton behaviour rule.

```
struct S { x : vec2<f32>; };
struct T { x : i32; };
[[group(0),binding(0)]] var<storage> a : S;
[[group(0),binding(1)]] var<storage> b : S;
[[group(0),binding(2)]] var<storage> c : S;
fn bar(i : i32) -> vec2<f32> {
  if (gid < i) {
    return a.x;
  } else {
    return b.x;
  }
}
fn bar() {
  let x = bar(c.x);
  textureSample(t, s, x);
}
```

With the only special singleton behaviour being `Next` right now, `bar` will be analyzed as ControlFlowCannotBeUniformAfter, but the call to `bar` will have the behaviour `{Next}`. I believe the singleton special case applies to function calls so that means the texture statement will be ok since its control flow start node would bypass the call to `bar`. This produces an inconsistency in the analysis. Control flow cannot be converged for `bar`, but the analysis determines the texture statement in `foo` is ok. 

This is related to #1770 in that we haven't fully settled on whether we want to spec functions as reconvergent.